SM-1082: Add SM Go SDK Wrapper

coltonhurst opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

<!--Describe what the purpose of this PR is. For example: what bug you're fixing or what new feature you're adding-->

Add the ability to easily use SM state in Go to the `bitwarden_client.go`.

## Code changes

<!--Explain the changes you've made to each file or major component. This should help the reviewer understand your changes-->
<!--Also refer to any related changes or PRs in other repositories-->

- **bitwarden_client.go:** Add the `statePath` parameter
- **example.go:** Set state in the example

## Before you submit

- Please add **unit tests** where it makes sense to do so

Thomas-Avery commented on languages/go/example/example.go:
```diff
@@ -16,14 +16,15 @@ func main() {
 	}
 
-	err := bitwardenClient.AccessTokenLogin(accessToken)
+	err := bitwardenClient.AccessTokenLogin(accessToken, &statePath)
```
When not using state, is the idea to pass `nil` or a pointer to an empty string? 

```
err := bitwardenClient.AccessTokenLogin(accessToken, nil)
```

or

```
statePath := ""
err := bitwardenClient.AccessTokenLogin(accessToken, &statePath)
```

Does it work the same way regardless?

coltonhurst replied:
For a developer using this function, the end result is the same either way. The correct thing to do is pass `nil`, as the underlying function in Rust takes an `Option`, and `nil` maps to the variant `None`. However, in the case an empty string is passed, the SDK won't be able to read or write to a state file, so it ends up being the same thing.

Thomas-Avery reviewed (approved):
LGTM

dani-garcia commented on languages/go/bitwarden_client.go:
```diff
@@ -45,8 +45,8 @@ func NewBitwardenClient(apiURL *string, identityURL *string) (*BitwardenClient,
 
-func (c *BitwardenClient) AccessTokenLogin(accessToken string) error {
-	req := AccessTokenLoginRequest{AccessToken: accessToken}
+func (c *BitwardenClient) AccessTokenLogin(accessToken string, statePath *string) error {
```
Nit: We're calling this new variable here `statePath`. In Python we've called it `state_file_path`. In Rust it's just `state_file`.

Should we decide on a consistent naming for them all?

coltonhurst replied:
Yes, that's a great point. I think `statePath` is fairly clear, any thoughts on that?

I have a draft PR here for the other places in the SDK & in the Python wrapper: https://github.com/bitwarden/sdk/pull/587

dani-garcia replied:
Yeah, that works for me 👍

dani-garcia commented on languages/go/example/example.go:
```diff
@@ -16,14 +16,15 @@ func main() {
 	bitwardenClient, _ := sdk.NewBitwardenClient(&apiURL, &identityURL)
 
 	accessToken := os.Getenv("ACCESS_TOKEN")
+	statePath := os.Getenv("STATE_PATH")
```
We should mention that this is optional, like we do with the API and identity URLs

coltonhurst replied:
Updated!

dani-garcia reviewed (commented):
We should make sure to update the README.md in the Go folder to include the new parameter to AccessTokenLogin too.

dani-garcia commented on languages/go/example/example.go:
```diff
@@ -19,11 +19,15 @@ func main() {
 	projectName := os.Getenv("PROJECT_NAME")
 
+	// Configuring the statePath is optional, pass nil
+	// in AccessTokenLogni() to not use state
```
```suggestion
	// in AccessTokenLogin() to not use state
```

coltonhurst replied:
Oh, that's terrible 🤦‍♂️ Fixed, sorry about that!

dani-garcia reviewed (commented):
We're missing a reference to the old AccessTokenLogin without statepath in the Readme: 

https://github.com/bitwarden/sdk/blob/1595306d0f80807983c374aaebc4938a835be78c/languages/go/README.md?plain=1#L42

Also it still has the old return type, while now it just returns an error, so we should change that too

coltonhurst commented:
> We're missing a reference to the old AccessTokenLogin without statepath in the Readme:
> 
> https://github.com/bitwarden/sdk/blob/1595306d0f80807983c374aaebc4938a835be78c/languages/go/README.md?plain=1#L42
> 
> Also it still has the old return type, while now it just returns an error, so we should change that too

Good call. I've updated the README and added a very small description regarding using `statePath` or `nil`.

differsthecat commented on languages/go/README.md:
```diff
@@ -36,10 +36,12 @@ bitwardenClient, _ := sdk.NewBitwardenClient(&apiURL, &identityURL)
 ### Login
 
-To login using an access token:
+To login using an access token. Define some `statePath` and pass it to use state, or pass `nil` instead to not use state.
```
Colton is out, I'm going to merge this PR now to get it in the hands of our vendor and and create a separate PR with these comment updates.
